cmd/network-console-collector/internal/collector/collector.go

```diff
@@ -0,0 +1,410 @@
+	c.session.Start(ctx)
+	g, ctx := errgroup.WithContext(ctx)
+	g.Go(c.runSession(ctx))
+	g.Go(c.runWorkQueue(ctx))
+	g.Go(c.monitoring(ctx))
+	g.Go(c.runDiscovery(ctx))
+	g.Go(c.runRecordCleanup(ctx))
+	g.Go(c.processManager.run(ctx))
+	g.Go(c.addressManager.run(ctx))
+	g.Go(c.pairManager.run(ctx))
+	return g.Wait()
+}
+
+func (c *Collector) updateGraph(event changeEvent, stor readonly) {
+	if dEvent, ok := event.(deleteEvent); ok {
+		c.graph.Unindex(dEvent.Record)
+		return
+	}
+	entry, ok := stor.Get(event.ID())
+	if !ok {
+		return
+	}
+	c.graph.Reindex(entry.Record)
+}
+
+func (c *Collector) monitoring(ctx context.Context) func() error {
+	eventQueueSpace := c.metrics.internal.queueUtilization.WithLabelValues("records")
+	return func() error {
+		defer func() {
+			c.logger.Info("collector monitoring shutdown complete")
+		}()
+		recordsCapacity := float64(cap(c.events))
+
+		utilization := func(queue chan changeEvent, capacity float64) float64 {
+			events := float64(len(queue))
+			return events / capacity
+		}
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case <-time.After(time.Second * 5):
+				recordsUtil := utilization(c.events, recordsCapacity)
+				eventQueueSpace.Set(recordsUtil)
+			}
+		}
+	}
+}
+
+func (c *Collector) runWorkQueue(ctx context.Context) func() error {
+	// reactors respond to record changes. Should be quick, and perform minimal
+	// read-only store ops. Handle any changes out of band.
+	reactors := map[vanflow.TypeMeta][]func(event changeEvent, stor readonly){}
+	for _, r := range standardRecordTypes {
+		reactors[r] = append(reactors[r], c.updateGraph)
+	}
+
+	reactors[AddressRecord{}.GetTypeMeta()] = append(reactors[AddressRecord{}.GetTypeMeta()], c.updateGraph)
+	reactors[FlowSourceRecord{}.GetTypeMeta()] = append(reactors[FlowSourceRecord{}.GetTypeMeta()], c.processManager.handleChangeEvent)
+	reactors[vanflow.ListenerRecord{}.GetTypeMeta()] = append(reactors[vanflow.ListenerRecord{}.GetTypeMeta()], c.addressManager.handleChangeEvent)
+	reactors[vanflow.ConnectorRecord{}.GetTypeMeta()] = append(reactors[vanflow.ConnectorRecord{}.GetTypeMeta()],
+		c.addressManager.handleChangeEvent,
+		c.processManager.handleChangeEvent,
+	)
+	reactors[vanflow.ProcessRecord{}.GetTypeMeta()] = append(reactors[vanflow.ProcessRecord{}.GetTypeMeta()],
+		c.processManager.handleChangeEvent,
+	)
+	reactors[ProcPairRecord{}.GetTypeMeta()] = append(reactors[ProcPairRecord{}.GetTypeMeta()], c.pairManager.handleChangeEvent)
+
+	return func() error {
+		defer func() {
+			c.logger.Info("queue worker shutdown complete")
+		}()
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case event := <-c.events:
+				start := time.Now()
+				typ := event.GetTypeMeta()
+				event.ID()
+				for _, reactor := range reactors[typ] {
+					reactor(event, c.Records)
+				}
+				c.metrics.internal.flowProcessingTime.WithLabelValues(typ.String()).Observe(time.Since(start).Seconds())
```

Comment:
TypeMeta.APIVersion is currently set to v1alpha1. If we eventually change versions to v2alpha1, for example,
and lets say we have two sites on the same VAN, one providing v1alpha1 and the other v2alpha1. Would that
cause metrics to be incompatible between these two sites?

Reply:
Interesting question! I'm not sure about "compatibility" - they would become two separate time series (i.e. skupper_internal_processing_seconds{type=v1alpha1.RouterRecord} from collector _old_ and {type=v2alpha.SiteRecord} from collector _new_.) The crux, I think, would be any dashboards or monitors/alerts for a particular label value would need updated.